diff --git a/ksonnet/dkube/all.libsonnet b/ksonnet/dkube/all.libsonnet
index 42884b3..9cfe972 100644
--- a/ksonnet/dkube/all.libsonnet
+++ b/ksonnet/dkube/all.libsonnet
@@ -2,9 +2,11 @@
   parts(params):: {
     local daemonset = import "dkube/dkube/daemonset.libsonnet",
     local deployment = import "dkube/dkube/deployment.libsonnet",
+    local statefulset = import "dkube/dkube/statefulset.libsonnet",
 
     all:: daemonset.all(params)
           + deployment.all(params)
+          + statefulset.all(params)
   },
 
 }
diff --git a/ksonnet/dkube/deployment.libsonnet b/ksonnet/dkube/deployment.libsonnet
index 34c23f4..c60c1c6 100644
--- a/ksonnet/dkube/deployment.libsonnet
+++ b/ksonnet/dkube/deployment.libsonnet
@@ -2,7 +2,6 @@
     all(params):: [
 	$.parts(params.namespace).logstash(params.tag, params.logstashImage, params.dkubeDockerSecret),
 	$.parts(params.namespace).dkubeEtcd(params.tag, params.dkubePVC),
-	$.parts(params.namespace).dkubeD3api(params.tag, params.dkubeApiServerImage, params.dkubeApiServerAddr, params.dkubeMountPath, params.dkubeApiServerAddr, params.rdmaEnabled, params.dkubeDockerSecret, params.minioSecretKey),
 	$.parts(params.namespace).dfabProxy(params.tag,params.dfabProxyImage, params.dkubeDockerSecret),
 	$.parts(params.namespace).ambassdor(params.tag),
     ],
@@ -140,144 +139,6 @@
 		}
 	    }
 	},
-
-	dkubeD3api(tag, apiServerImage, apiServerAddr, mountPath, dkubeApiServerAddr, isRdmaEnabled, dkubeDockerSecret, minioSecretKey):: {
-	    local dkubeApiServerAddrArray = std.split(dkubeApiServerAddr, ":"),
-	    local dkubeApiServerPort = std.parseInt(dkubeApiServerAddrArray[std.length(dkubeApiServerAddrArray)-1]),
-
-	    "apiVersion": "extensions/v1beta1", 
-	    "kind": "Deployment", 
-	    "metadata": {
-		"labels": {
-		    "app": "dkube-d3api"
-		}, 
-		"name": "dkube-d3api-" + tag , 
-		"namespace": namespace
-	    }, 
-	    "spec": {
-		"selector": {
-		    "matchLabels": {
-			"app": "dkube-d3api"
-		    }
-		}, 
-		"template": {
-		    "metadata": {
-			"labels": {
-			    "app": "dkube-d3api"
-			}
-		    }, 
-		    "spec": {
-			"imagePullSecrets": [
-			{
-			    "name": dkubeDockerSecret
-			}
-			],
-			"nodeSelector": {
-				"d3.nodetype": "dkube"
-			},
-			"tolerations": [
-				{
-					"operator": "Exists"
-				},
-			],
-			"containers": [
-			{
-			    "image": apiServerImage, 
-			    "imagePullPolicy": "IfNotPresent", 
-			    "name": "dkube-d3api", 
-			    "securityContext": {
-				"runAsUser": 0
-			    },
-			    "ports": [
-			    {
-				"containerPort": dkubeApiServerPort, 
-				"name": "dkube-d3api", 
-				"protocol": "TCP"
-			    }
-			    ], 
-			    "env": [
-			    {
-				"name": "DKUBE_MOUNT_PATH", 
-				"value": mountPath
-			    },
-			    {
-				"name": "DKUBE_SERVICE_ACCOUNT",
-				"value": "dkube"
-			    },
-			    {
-				"name": "RDMA_ENABLED",
-				"value": std.toString(isRdmaEnabled)
-			    }
-			    ], 
-			    "volumeMounts": [
-			    {
-				"mountPath": mountPath, 
-				"name": "store"
-			    },
-			    {
-				"mountPath": "/var/log/minio/dkube",
-				"name": "dkube-logs"
-			    },
-				{
-				"mountPath": "/var/log/dkube",
-				"name": "dkube-logs-host"
-				}
-			    ]
-			}
-			], 
-            "dnsConfig": {
-                "options": [
-                    {
-                        "name": "single-request-reopen"
-                    },
-                    {
-                        "name": "timeout",
-                        "value": "30"
-                    }
-                ]
-            },
-			"serviceAccount": "dkube", 
-			"volumes": [
-			{
-			    "flexVolume": {
-				"driver": "oc/d3", 
-				"options": {
-				    "accessKey": "dkube", 
-				    "bucket": "dkube", 
-				    "endpoint": "http://127.0.0.1:32223",
-				    "s3provider": "minio", 
-				    "secretKey": minioSecretKey
-				}
-			    }, 
-			    "name": "store"
-			},
-			{
-			    "flexVolume": {
-				"driver": "oc/d3", 
-				"options": {
-				    "accessKey": "dkube", 
-				    "bucket": "logs", 
-                    "prefix": "dkube",
-				    "endpoint": "http://127.0.0.1:32223",
-				    "s3provider": "minio", 
-				    "secretKey": minioSecretKey
-				}
-			    }, 
-			    "name": "dkube-logs"
-			},
-			{
-				"hostPath": {
-					"path": "/var/log/dkube",
-					"type": "DirectoryOrCreate"
-				},
-				"name": "dkube-logs-host"
-			}
-			]
-		    }
-		}
-	    }
-	}, 
-
 	dfabProxy(tag , dfabProxyImage, dkubeDockerSecret):: {
 	    "apiVersion": "extensions/v1beta1",
 	    "kind": "Deployment",
diff --git a/ksonnet/service/dkube.libsonnet b/ksonnet/service/dkube.libsonnet
index 2eb1695..049c01a 100644
--- a/ksonnet/service/dkube.libsonnet
+++ b/ksonnet/service/dkube.libsonnet
@@ -99,7 +99,35 @@
         "namespace": namespace
       }, 
       "spec": {
-        "clusterIP": "None", 
+        "ports": [
+          {
+            "name": "dkube-d3api", 
+            "port": dkubeApiServerPort, 
+            "protocol": "TCP", 
+            "targetPort": dkubeApiServerPort
+          }
+        ], 
+        "selector": {
+          "app": "dkube-d3api"
+        }, 
+        "type": "ClusterIP"
+      }
+    }, //service
+    dkubeService(dkubeApiServerAddr):: {
+      local dkubeApiServerAddrArray = std.split(dkubeApiServerAddr, ":"),
+      local dkubeApiServerPort = std.parseInt(dkubeApiServerAddrArray[std.length(dkubeApiServerAddrArray)-1]),
+
+      "apiVersion": "v1", 
+      "kind": "Service", 
+      "metadata": {
+        "labels": {
+          "app": "dkube-d3api"
+        }, 
+        "name": "dkube-d3api-headless", 
+        "namespace": namespace
+      }, 
+      "spec": {
+        "clusterIP": "None",
         "ports": [
           {
             "name": "dkube-d3api", 
